dll/win32/comctl32/edit.c

```diff
@@ -0,0 +1,5040 @@
+				line_idx--;
+			}
+			w = es->format_rect.right - es->format_rect.left;
+			lw = line_def->width;
+
+			if (es->style & ES_RIGHT)
+				x = w - (lw - x);
+			else if (es->style & ES_CENTER)
+				x += (w - lw) / 2;
+		}
+		x += es->format_rect.left;
+	}
+
+	if (rev)
+	{
+		li = EDIT_EM_LineIndex(es, line);
+		ll = EDIT_EM_LineLength(es, li);
+		s = min(es->selection_start, es->selection_end);
+		e = max(es->selection_start, es->selection_end);
+		s = min(li + ll, max(li, s));
+		e = min(li + ll, max(li, e));
+	}
+
+	if (ssa)
+		ScriptStringOut(ssa, x, y, 0, &es->format_rect, s - li, e - li, FALSE);
+	else if (rev && (s != e) &&
+			((es->flags & EF_FOCUSED) || (es->style & ES_NOHIDESEL))) {
+		x += EDIT_PaintText(es, dc, x, y, line, 0, s - li, FALSE);
+		x += EDIT_PaintText(es, dc, x, y, line, s - li, e - s, TRUE);
+		x += EDIT_PaintText(es, dc, x, y, line, e - li, li + ll - e, FALSE);
+	} else
+		x += EDIT_PaintText(es, dc, x, y, line, 0, ll, FALSE);
+}
+
+
+/*********************************************************************
+ *
+ *	EDIT_AdjustFormatRect
+ *
+ *	Adjusts the format rectangle for the current font and the
+ *	current client rectangle.
+ *
+ */
+static void EDIT_AdjustFormatRect(EDITSTATE *es)
+{
+	RECT ClientRect;
+
+	es->format_rect.right = max(es->format_rect.right, es->format_rect.left + es->char_width);
+	if (es->style & ES_MULTILINE)
+	{
+	    INT fw, vlc, max_x_offset, max_y_offset;
+
+	    vlc = get_vertical_line_count(es);
+	    es->format_rect.bottom = es->format_rect.top + vlc * es->line_height;
+
+	    /* correct es->x_offset */
+	    fw = es->format_rect.right - es->format_rect.left;
+	    max_x_offset = es->text_width - fw;
+	    if(max_x_offset < 0) max_x_offset = 0;
+	    if(es->x_offset > max_x_offset)
+		es->x_offset = max_x_offset;
+
+	    /* correct es->y_offset */
+	    max_y_offset = es->line_count - vlc;
+	    if(max_y_offset < 0) max_y_offset = 0;
+	    if(es->y_offset > max_y_offset)
+		es->y_offset = max_y_offset;
+
+	    /* force scroll info update */
+	    EDIT_UpdateScrollInfo(es);
+	}
+	else
+	/* Windows doesn't care to fix text placement for SL controls */
+		es->format_rect.bottom = es->format_rect.top + es->line_height;
+
+	/* Always stay within the client area */
+	GetClientRect(es->hwndSelf, &ClientRect);
+	es->format_rect.bottom = min(es->format_rect.bottom, ClientRect.bottom);
+
+	if ((es->style & ES_MULTILINE) && !(es->style & ES_AUTOHSCROLL))
+		EDIT_BuildLineDefs_ML(es, 0, get_text_length(es), 0, NULL);
+
+	EDIT_SetCaretPos(es, es->selection_end, es->flags & EF_AFTER_WRAP);
+}
+
+
+/*********************************************************************
+ *
+ *	EDIT_SetRectNP
+ *
+ *	note:	this is not (exactly) the handler called on EM_SETRECTNP
+ *		it is also used to set the rect of a single line control
+ *
+ */
+static void EDIT_SetRectNP(EDITSTATE *es, const RECT *rc)
+{
+	LONG_PTR ExStyle;
+	INT bw, bh;
+	ExStyle = GetWindowLongPtrW(es->hwndSelf, GWL_EXSTYLE);
```

Comment:
Sometimes `GWL_EXSTYLE` is with Ptr, sometimes without Ptr.
Which one is correct?

Reply:
GWLP when using GetWindowLongPtr.

Reply:
Note that GWL_* and GWLP_* have the same values, but if calling GetWindowLongPtr, the GWLP_ defines should always be used, in case future flags somehow diverged.

Reply:
Do you guys have any idea how to fix it properly and avoid tons of diff with wine?

Reply:
Well.... send upstream is the only idea I have that would properly solve it.